src/tty/unix.rs

```diff
@@ -15,59 +15,79 @@
 //! tty related functionality
 //!
 
-use crate::tty::EventedReadWrite;
+use crate::tty::{EventedReadWrite, ChildInfo};
 use crate::term::SizeInfo;
 use crate::display::OnResize;
 use crate::config::{Config, Shell};
 use crate::cli::Options;
 use mio;
 
-use libc::{self, c_int, pid_t, winsize, SIGCHLD, TIOCSCTTY, WNOHANG};
+use libc::{self, c_int, winsize, SIGCHLD, TIOCSCTTY};
 
-use std::os::unix::io::{FromRawFd, RawFd};
+use std::os::unix::io::{FromRawFd};
 use std::fs::File;
 use std::os::unix::process::CommandExt;
-use std::process::{Command, Stdio};
+use std::process::{Command, Stdio, Child};
 use std::ffi::CStr;
 use std::ptr;
 use mio::unix::EventedFd;
-use std::io;
+use std::io::{self, Write, Read};
 use std::os::unix::io::AsRawFd;
+use std::mem;
 
+/// Get the current value of errno
+fn errno() -> c_int {
+    ::errno::errno().0
+}
 
-/// Process ID of child process
-///
-/// Necessary to put this in static storage for `sigchld` to have access
-pub static mut PID: pid_t = 0;
+fn set_errno(value: c_int) {
+    ::errno::set_errno(::errno::Errno(value));
+}
 
-/// Exit flag
+/// Self-pipe
 ///
-/// Calling exit() in the SIGCHLD handler sometimes causes opengl to deadlock,
-/// and the process hangs. Instead, this flag is set, and its status can be
-/// checked via `process_should_exit`.
-static mut SHOULD_EXIT: bool = false;
-
-extern "C" fn sigchld(_a: c_int) {
-    let mut status: c_int = 0;
-    unsafe {
-        let p = libc::waitpid(PID, &mut status, WNOHANG);
-        if p < 0 {
-            die!("Waiting for pid {} failed: {}\n", PID, errno());
+/// We write any signals into the pipe, and the event loop wakes up and reads them.
+static mut SIG_PIPE_WRITE: Option<File> = None;
+
+unsafe extern "C" fn on_signal(sig: c_int) {
+    let old_errno = errno();
+
+    let byte = [sig as u8];
+    let file: &mut File = SIG_PIPE_WRITE.as_mut().unwrap();
+    match file.write_all(&byte) {
+        Ok(_) => {
+            // TODO: a flush is almost certainly unnecesary
+            file.flush().unwrap();
+            // re-register
+            libc::signal(sig, on_signal as _);
         }
-
-        if PID == p {
-            SHOULD_EXIT = true;
+        Err(e) => {
+            // if we get EPIPE, it's just because we're shutting down anyway, so ignore it
+            if e.kind() != io::ErrorKind::BrokenPipe {
+                die!("Writing signal {} to self-pipe failed: {}\n", sig, e);
+            }
         }
     }
-}
 
-pub fn process_should_exit() -> bool {
-    unsafe { SHOULD_EXIT }
+    set_errno(old_errno);
 }
 
-/// Get the current value of errno
-fn errno() -> c_int {
-    ::errno::errno().0
+// lifted from nix::unistd (MIT)
+fn pipe() -> (File, File) {
+    unsafe {
+        let mut fds: [c_int; 2] = mem::uninitialized();
+        if let 0 = libc::pipe(fds.as_mut_ptr()) {
+            set_nonblocking(fds[0]);
+            set_nonblocking(fds[1]);
```

Comment:
Non-blocking wouldn't seem to have any benefit, but _everyone_ seems to do it with self-pipes. Let me know if it's unnecessary.